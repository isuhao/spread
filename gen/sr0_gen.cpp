#include "sr0_gen.hpp"
#include "listwriter.hpp"
#include <mangle/stream/servers/outfile_stream.hpp>
#include <boost/filesystem.hpp>

using namespace SpreadGen;
using namespace Spread;

void GenSR0::makeSR0(const std::string &arcFile, const std::string &outDir)
{ makeSR0(cache.index.addFile(arcFile), outDir); }

void GenSR0::makeSR0(const Hash &hash, const std::string &outDir)
{
  boost::filesystem::path out = outDir;

  PackLister lst(cache, rules);
  Hash dir = lst.addDir("index", hash);

  ListWriter wrt(cache);
  wrt.write(lst, (out / "_output").string());

  // Write first 8 bytes of hash to 'short.txt'
  Mangle::Stream::OutFileStream::Write
    ((out / "short.txt").string(), dir.toString().substr(0,8));
}
