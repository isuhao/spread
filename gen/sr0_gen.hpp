#ifndef __SPREAD_GEN_SR0GEN_HPP_
#define __SPREAD_GEN_SR0GEN_HPP_

#include "cache/cache.hpp"
#include "rules/ruleset.hpp"

namespace SpreadGen
{
  struct GenSR0
  {
    GenSR0(Cache::Cache &_cache,
           const Spread::RuleSet &_rules)
      : cache(_cache), rules(_rules) {}

    /* Create an SR0 output set. Consists of:

       short.txt - short version of output directory hash
       _output - directory to be zipped and uploaded as index.zip

       The input is an archive file that contains the full output
       directory. We assume that the RuleSet contains an ARC rule for
       the archive, and that the directory object is loadable through
       the Cache.
     */
    void makeSR0(const std::string &arcFile, const std::string &outDir);

    /*
      Version that takes an archive hash or dir hash as parameter,
      instead of an archive file.
     */
    void makeSR0(const Spread::Hash &hash, const std::string &outDir);

  private:
    Cache::Cache &cache;
    const Spread::RuleSet &rules;
  };
}

#endif
