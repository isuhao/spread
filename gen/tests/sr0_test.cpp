#include "sr0_gen.hpp"
#include <iostream>
#include "cache/index.hpp"

#include "print_dir.hpp"

using namespace SpreadGen;
using namespace Spread;
using namespace std;

Cache::CacheIndex cache;
RuleSet rules;

Hash dirHash;
Hash hello("hello",5);
Hash world("world",5);

Hash arcHash("ARC_WITH_DIR");

int main()
{
  Hash::DirMap dir;
  dir["file1.txt"] = hello;
  dir["file2.txt"] = world;
  dirHash = Dir::write(dir, "_sr0_dir.out");
  cache.addFile("_sr0_dir.out");

  rules.addURL(hello, "url-to-hello");
  rules.addURL(arcHash, "url-to-archive");
  rules.addArchive(arcHash, dirHash);

  GenSR0 gen(cache, rules);
  gen.makeSR0(dirHash, "_sr0_output");

  printDir("_sr0_output");

  return 0;
}
