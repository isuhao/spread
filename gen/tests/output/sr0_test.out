
Directory: _sr0_output
00                                                 _output
00                                                 _output/dirs
U0D_QYj52mHfrx2A__Dtg4D3hiKo29FBdTunQM-y6e9u       _output/dirs/U0D_QYj52mHfrx2A__Dtg4D3hiKo29FBdTunQM-y6e9u
5sKBIks5Q1SHRJZlyw6IMcQUMlW54LedwwaYzKVVYjlV       _output/packs.json
GRF6eD3ilJ4kxqhNO2gJjQCpYjOe6z5ZCKV4wzxLQSSs       _output/rules.json
rw6jJOkwZ36fhHadDma5d6fbbDEpt-HZexJgRgGBdS3xAw     index.zip
PtuyGzC1hSLbx-4y8p0C3PsrZbGnGz35oXzMSwTkFDUI       short.txt
Total 7 elements
Hash: XK20cLOtLaXFTv7d0Oy0lgU13DRbyHVMrj8H_-c0fqOwAQ
